### PR TITLE
API changes for client-initiated termination reason

### PR DESCRIPTION
**Pull request by yguo0905**
**What this PR does / why we need it**:

This PR contains the API changes for https://github.com/kubernetes/community/pull/541. 

**Special notes for your reviewer**:

The logic that generates and delivers the termination reason in kubectl and kubelet will be in separate PRs. 

**Release note**:
```
NONE
```

### CODE REVIEW

**Comment by yguo0905** [comment]
@kow3ns

File: pkg/api/defaulting_test.go [hunk 1]
```diff
@@ -160,7 +160,7 @@ func TestDefaulting(t *testing.T) {
 				if !expectedChanged || changedOnce {
 					break
 				}
-				if iter > 300 {
+				if iter > 500 {
```
Comment (ncdc) [inline pkg/api/defaulting_test.go]: Why this change?

File: pkg/api/types.go [hunk 2]
```diff
@@ -1595,6 +1595,45 @@ type Handler struct {
 	TCPSocket *TCPSocketAction
 }
 
+// DeleteExecAction describes a "run in container" action that will be invoked
+// inside of a container prior to sending the TERM signal to the container's
```
Comment (ncdc) [inline pkg/api/types.go]: Note that now that https://github.com/kubernetes/kubernetes/pull/45236 has merged, the PID namespace is shared by default. I don't know that saying it sends TERM to the container's entrypoint makes sense any more. WDYT?

File: pkg/api/types.go [hunk 3]
```diff
@@ -1595,6 +1595,45 @@ type Handler struct {
 	TCPSocket *TCPSocketAction
 }
 
+// DeleteExecAction describes a "run in container" action that will be invoked
```
Comment (ncdc) [inline pkg/api/types.go]: Since the type is DeleteExecAction do you want to say that it describes an "exec in container" action?

File: pkg/api/types.go [hunk 4]
```diff
@@ -1595,6 +1595,45 @@ type Handler struct {
...
+type DeleteHTTPGetAction struct {
+	HTTPGetAction
+
+	// ReasonHeader is the header that will be set to the reason for a
+	// deletion.
+	// +optional
+	ReasonHeader string
+}
+
+// PreStopHandler invokes either a DeleteExecAction, a DeleteHTTPGetAction or a
```
Comment (ncdc) [inline pkg/api/types.go]: comma after DeleteHTTPGetAction

File: pkg/api/types.go [hunk 5]
```diff
@@ -1595,6 +1595,45 @@ type Handler struct {
...
+// TCPSocketAction prior to the graceful termination of a Pod. One and only one
+// of the fields should be specified.
```
Comment (ncdc) [inline pkg/api/types.go]: s/should/may/

File: pkg/api/v1/meta.go [hunk 6]
```diff
@@ -51,6 +51,9 @@ func (meta *ObjectMeta) GetDeletionGracePeriodSeconds() *int64 { return meta.Del
 func (meta *ObjectMeta) SetDeletionGracePeriodSeconds(deletionGracePeriodSeconds *int64) {
 	meta.DeletionGracePeriodSeconds = deletionGracePeriodSeconds
 }
+func (meta *ObjectMeta) GetDeletionReason() string { return "" }
```
Comment (ncdc) [inline pkg/api/v1/meta.go]: Are these supposed to be left unimplemented?

File: pkg/api/validation/validation.go [hunk 7]
```diff
@@ -1802,13 +1830,38 @@ func validateHandler(handler *api.Handler, fldPath *field.Path) field.ErrorList
 	return allErrors
 }
 
+func validatePreStopHandler(handler *api.PreStopHandler, fldPath *field.Path) field.ErrorList {
+	numHandlers := 0
+	allErrors := field.ErrorList{}
+	if handler.Exec != nil {
+		if numHandlers > 0 {
+			allErrors = append(allErrors, field.Forbidden(fldPath.Child("exec"), "may not specify more than 1 handler type"))
```
Comment (ncdc) [inline pkg/api/validation/validation.go]: Not sure how you'd ever hit this

File: pkg/kubelet/pleg/generic.go [hunk 8]
```diff
@@ -186,8 +186,6 @@ func (g *GenericPLEG) relist() {
 	}
 
 	timestamp := g.clock.Now()
-	// Update the relist time.
```
Comment (ncdc) [inline pkg/kubelet/pleg/generic.go]: Why did you move this?

**Comment by ncdc** [comment]
@derekwaynecarr @yujuhong @Random-Liu

File: pkg/api/defaulting_test.go (hunk 1 above)
Comment (yguo0905) [inline pkg/api/defaulting_test.go]: The test generated 300 objects, which were enough to cover all the fields with default values. Since this change added nested fields to `Lifecycle`, we now need more permutations to be generated for the test objects in order to ensure that the additional fields with default values are covered. This happened before in https://github.com/kubernetes/kubernetes/pull/39181. I will create a separate issue for a real fix.

File: pkg/api/types.go (hunk 2 above)
Comment (yguo0905) [inline pkg/api/types.go]: Can you please elaborate on what the shared PID namespace implies here? How does it relate to the termination procedure?

File: pkg/api/types.go (hunk 3 above)
Comment (yguo0905) [inline pkg/api/types.go]: The `Delete` indicates it's an `ExecAction` plus a deletion reason string. Made it clear in the comment.

File: pkg/api/v1/meta.go (hunk 6 above)
Comment (yguo0905) [inline pkg/api/v1/meta.go]: I think so. The `ObjectMeta` in `pkg/api/v1/meta.go` is deprecated and we add the two function here to make it satisfy the interface.

File: pkg/kubelet/pleg/generic.go (hunk 8 above)
Comment (yguo0905) [inline pkg/kubelet/pleg/generic.go]: I think this is a mistake introduced during rebase.

File: pkg/api/types.go (hunk 2 above)
Comment (ncdc) [inline pkg/api/types.go]: It means that if you have a pod with multiple containers, the PID namespace is shared among all the containers. Before this change, each container's PID 1 was its entrypoint, and if you killed PID 1 in a single container, that container would die, but the other containers in the pod would be unaffected. With a shared PID namespace, PID 1 is owned by the pod sandbox (is that the right terminology @derekwaynecarr?), and each container's entrypoint has some unique PID other than 1. If you kill PID 1 in the shared PID namespace, you will bring down all the containers for that pod.

Comment (yguo0905) [inline pkg/api/types.go]: Thanks. So the issue here is the "entrypoint"? Removed it in the latest commit.

File: pkg/api/types.go (hunk 4 above)
Comment (yguo0905) [inline pkg/api/types.go]: Done.

File: pkg/api/types.go (hunk 5 above)
Comment (yguo0905) [inline pkg/api/types.go]: Done.

File: pkg/api/validation/validation.go (hunk 7 above)
Comment (yguo0905) [inline pkg/api/validation/validation.go]: Removed this condition.

**Comment by yguo0905** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by dims** [comment]
@yguo0905 Do we need this for 1.9? please rebase if so.

**Comment by derekwaynecarr** [comment]
/do-not-merge
just noting that the community design pr is still not resolved.
